lace_ground.py

```diff
@@ -13,215 +13,274 @@
-					line = line.lstrip('[')
-					line = line.rstrip(']\t\n')
-					rowData = line.split(']\t[')
-					data.append([])
-					for cell in rowData:
-						data[-1].append([float(num) for num in cell.split(',')])
-						
-		return {"type":type, "rowCount":rowCount, "colCount":colCount, "data":data}
-			
-	def drawCheckerGround(self, data, rowCount, colCount):
-		a = self.options.spacing
-		theta = radians(self.options.angle)
-		deltaX = a*sin(theta) 
-		deltaY = a*cos(theta)
-		maxRows = ceil(self.options.height / deltaY)
-		maxCols = ceil(self.options.width  / deltaX)
-		
-		x = 0.0
-		y = 0.0
-		repeatY = 0
-		repeatX = 0
-
-		while repeatY * rowCount < maxRows:
-			x = 0.0
-			repeatX = 0
-			
-			while repeatX * colCount < maxCols:
-				
-				for row in data:
-					for coords in row:
-						x1 = x + coords[0]*deltaX
-						y1 = y + coords[1]*deltaY
-						x2 = x + coords[2]*deltaX
-						y2 = y + coords[3]*deltaY
-						x3 = x + coords[4]*deltaX
-						y3 = y + coords[5]*deltaY
-				
-						self.line(x1,y1,x2,y2)
-						self.line(x1,y1,x3,y3)
-					
-				repeatX += 1
-				x += deltaX * colCount
-
-			repeatY += 1
-			y += deltaY * rowCount
-
-	def unsignedLong(self, signedLongString):
-		longColor = long(signedLongString)
-		if longColor < 0:
-			longColor = longColor & 0xFFFFFFFF
-		return longColor
-
-	def getColorString(self, longColor):
-		"""
-		Convert numeric color value to hex string using formula A*256^0 + B*256^1 + G*256^2 + R*256^3
-		From: http://www.hoboes.com/Mimsy/hacks/parsing-and-setting-colors-inkscape-extensions/
-		"""
-		longColor = self.unsignedLong(longColor)
-		hexColor = hex(longColor)[2:-3]
-		hexColor = hexColor.rjust(6, '0')
-		return '#' + hexColor.upper()
-	
-	def effect(self):
-		"""
-		Effect behaviour.
-		Overrides base class' method and draws something.
-		"""
-		# Locate and load the file containing the lace ground descriptor
-		if self.fname == None:
-			self.fname = self.options.file
-			
-		if self.fname == '': sys.exit(1)
-		elif not path.isfile(self.fname): sys.exit(1)
-		
-		result = self.loadFile(self.fname)
-		
-		#Convert input from mm to pixels, assuming 90 dpi
-		conversion = 90.0 / 25.4
-		self.options.width *= conversion
-		self.options.height *= conversion
-		self.options.size *= conversion
-		
-		# Users expect spacing to be the vertical distance between footside pins (vertical distance between every other row) 
-		# but in the script we use it as as diagonal distance between grid points
-		# therefore convert spacing based on the angle chosen
-		theta = radians(self.options.angle)
-		self.options.spacing *= conversion/(2.0*cos(theta))
-		
-		# Convert color from long integer to hexidecimal string
-		self.options.color = self.getColorString(self.options.color)
-		
-		# Draw a ground based on file description and user inputs
-		# For now, assume style is Checker but could change in future
-		self.drawCheckerGround(result["data"],result["rowCount"],result["colCount"])
+        p1, p2, p3 = pairs

```

Comment:
Not sure yet what you mean here, still have to install your version try it, but I'm afraid you need some more background information about diagrams and patterns used by lace makers. A little is explained on https://github.com/d-bl/GroundForge/wiki#purpose-of-the-diagrams but still assumes advanced lace making experience. This plugin is dealing with the offline generated diagrams d-h, which are quite unorthodox for lace makers. I now see I need to elaborate/illustrate the last dot of the section. Please have some patience.

I suppose most of your changes are about issue #2 but this changes functionality, try to focus on single issues to simplify reviewing for me.

Try not to meddle too much which functionality anyway as without lace making experience you might mistake the purpose, though I'd welcome some way to simplify the nudging process explained at https://github.com/d-bl/GroundForge/wiki/Reshape-Patterns Not sure this dialog-based plugin mechanism can do anything about it. Start an new issue if you have ideas.